src/core/console_user_server/include/console_user_server/software_function_handler.hpp

```diff
@@ -62,17 +62,36 @@ class software_function_handler final : public pqrs::dispatcher::extra::dispatch
   }
 
   void execute_set_mouse_cursor_position(const software_function_details::set_mouse_cursor_position& set_mouse_cursor_position) {
+    if (auto target_display_id = get_target_display_id(set_mouse_cursor_position)) {
+      auto local_display_point = set_mouse_cursor_position.get_point(CGDisplayBounds(*target_display_id));
+      CGDisplayMoveCursorToPoint(*target_display_id, local_display_point);
+    }
+  }
+
+  std::optional<CGDirectDisplayID> get_target_display_id(const software_function_details::set_mouse_cursor_position& set_mouse_cursor_position) {
     if (auto screen = set_mouse_cursor_position.get_screen()) {
       auto active_displays = pqrs::osx::cg_display::active_displays();
       if (*screen < active_displays.size()) {
-        CGDisplayMoveCursorToPoint(active_displays[*screen],
-                                   set_mouse_cursor_position.get_point(
-                                       CGDisplayBounds(active_displays[*screen])));
+        return std::optional<CGDirectDisplayID>(active_displays[*screen]);
       }
     } else {
-      CGWarpMouseCursorPosition(set_mouse_cursor_position.get_point(
-          CGDisplayBounds(CGMainDisplayID())));
+      return get_active_display_id();
```

Comment:
This is the new code path, if `get_screen` is absent then it invokes `get_active_display_id` to get the display the mouse cursor is on.